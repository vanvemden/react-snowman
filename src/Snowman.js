import React, { useState } from "react";
import "./Snowman.css";
import img0 from "./0.png";
import img1 from "./1.png";
import img2 from "./2.png";
import img3 from "./3.png";
import img4 from "./4.png";
import img5 from "./5.png";
import img6 from "./6.png";
import { randomWord, ENGLISH_WORDS } from "./words";

function Snowman({ maxWrong, images, words }) {
  /** by default, allow 6 guesses and use provided gallows images. */

  const [nWrong, updateNWrong] = useState(0);
  const [guessed, updateGuessed] = useState(new Set());
  const [answer, updateAnswer] = useState(words[0]);

  /** guessedWord: show current-state of word:
    if guessed letters are {a,p,e}, show "app_e" for "apple"
  */
  function guessedWord() {
    return answer
      .split("")
      .map(ltr => (guessed.has(ltr) ? ltr : "_"));
  }

  /** handleGuess: handle a guessed letter:
      - add to guessed letters
      - if not in answer, increase number-wrong guesses
  */
  function handleGuess(evt) {
    let ltr = evt.target.value;
    updateGuessed(g => {
      const newGuessed = new Set(g);
      newGuessed.add(ltr);
      return newGuessed;
    });

    updateNWrong(n => n + (answer.includes(ltr) ? 0 : 1));
  }

  /** generateButtons: return array of letter buttons to render */
  function generateButtons() {
    return "abcdefghijklmnopqrstuvwxyz".split("").map(ltr => (
      <button
        key={ltr}
        value={ltr}
        onClick={handleGuess}
        disabled={guessed.has(ltr)}
      >
        {ltr}
      </button>
    ));
  }

  function handleRestart() {
    updateNWrong(0);
    updateGuessed(new Set());
    updateAnswer(randomWord(ENGLISH_WORDS));
  }

  /** render: render game */
  return (
    <div className="Snowman">
      <img src={images[nWrong]} data-testid="snowman-image" />
      <p className="Snowman-wrongs">Number wrong: {nWrong}</p>
      <p className="Snowman-word">{guessedWord()}</p>
      {(nWrong < maxWrong) ?
        <p>{generateButtons()}</p> :
        <p>You lose. The expected anwer is {answer}.</p>
      }
      <button onClick={handleRestart}>Restart Game</button>
    </div>
  );
}

Snowman.defaultProps = {
  maxWrong: 6,
  images: [img0, img1, img2, img3, img4, img5, img6],
  words: [randomWord(ENGLISH_WORDS)]
};

export default Snowman;
