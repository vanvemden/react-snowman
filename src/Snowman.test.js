import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Snowman from "./Snowman";
import { ENGLISH_WORDS, randomWord } from "./words";

// beforeEach(function () {
//   Math.random = jest.fn(() => 0);

//   jest
//     .spyOn(Math, "random")
//     .mockReturnValueOnce(0);

// });

// afterEach(function () {
//   //  randomWord.mockRestore();
// });

it("renders without crashing", function () {
  render(<Snowman />);
});

it("matches the snapshot", function () {
  const { asFragment } = render(<Snowman />);
  expect(asFragment()).toMatchSnapshot();
});

it("changes the image to img1 after the first incorrect guess", function () {
  const { getByTestId, getByText } = render(<Snowman words={["apple"]} />);
  const button = getByText("z");
  fireEvent.click(button);
  // expect.stringContaining(string) matches the received value 
  // if it is a string that contains the exact expected string.
  expect(getByTestId("snowman-image").src).toEqual(expect.stringContaining('1.png'));
});

it("does not change image but reveals letters change after correct guess", function () {
  const { queryByText, getByTestId, getByText } = render(<Snowman words={["apple"]} />);
  const wordBefore = queryByText('_____');
  expect(wordBefore).toBeTruthy();

  const button = getByText("a");
  fireEvent.click(button);

  const wordAfter = queryByText('a____');
  expect(wordAfter).toBeTruthy();

  expect(getByTestId("snowman-image").src).toEqual(expect.stringContaining('0.png'));
});

it("game ends after max. number of wrong guesses", function () {
  const { asFragment, getByText } = render(<Snowman words={["apple"]} />);
  // how to test if defaultProps are not set (but set as default values)??
  const maxWrong = Snowman.defaultProps.maxWrong;
  const word = "apple";
  const chars = "abcdefghijklmopqrstuvwxyz".split('').filter(c => !word.includes(c)).slice(0, maxWrong);
  for (let char of chars) {
    let button = getByText(char);
    fireEvent.click(button);
  }
  expect(asFragment()).toMatchSnapshot();
});

it("basic test for restart; game restarts after clicking on Restart button", function () {
  const { queryByText, getByTestId, getByText } = render(<Snowman words={["apple"]} />);
  const button = getByText("z");
  fireEvent.click(button);
  expect(queryByText('Number wrong: 1')).toBeInTheDocument();
  expect(getByTestId("snowman-image").src).toEqual(expect.stringContaining('1.png'));

  const restartButton = getByText("Restart Game");
  fireEvent.click(restartButton);
  expect(queryByText('Number wrong: 0')).toBeInTheDocument();
  expect(getByTestId("snowman-image").src).toEqual(expect.stringContaining('0.png'));
})