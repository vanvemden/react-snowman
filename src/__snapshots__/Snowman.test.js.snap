// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game ends after max. number of wrong guesses 1`] = `
<DocumentFragment>
  <div
    class="Snowman"
  >
    <img
      data-testid="snowman-image"
      src="6.png"
    />
    <p
      class="Snowman-wrongs"
    >
      Number wrong: 6
    </p>
    <p
      class="Snowman-word"
    >
      _____
    </p>
    <p>
      You lose. The expected anwer is apple.
    </p>
    <button>
      Restart Game
    </button>
  </div>
</DocumentFragment>
`;

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="Snowman"
  >
    <img
      data-testid="snowman-image"
      src="0.png"
    />
    <p
      class="Snowman-wrongs"
    >
      Number wrong: 0
    </p>
    <p
      class="Snowman-word"
    >
      ______
    </p>
    <p>
      <button
        value="a"
      >
        a
      </button>
      <button
        value="b"
      >
        b
      </button>
      <button
        value="c"
      >
        c
      </button>
      <button
        value="d"
      >
        d
      </button>
      <button
        value="e"
      >
        e
      </button>
      <button
        value="f"
      >
        f
      </button>
      <button
        value="g"
      >
        g
      </button>
      <button
        value="h"
      >
        h
      </button>
      <button
        value="i"
      >
        i
      </button>
      <button
        value="j"
      >
        j
      </button>
      <button
        value="k"
      >
        k
      </button>
      <button
        value="l"
      >
        l
      </button>
      <button
        value="m"
      >
        m
      </button>
      <button
        value="n"
      >
        n
      </button>
      <button
        value="o"
      >
        o
      </button>
      <button
        value="p"
      >
        p
      </button>
      <button
        value="q"
      >
        q
      </button>
      <button
        value="r"
      >
        r
      </button>
      <button
        value="s"
      >
        s
      </button>
      <button
        value="t"
      >
        t
      </button>
      <button
        value="u"
      >
        u
      </button>
      <button
        value="v"
      >
        v
      </button>
      <button
        value="w"
      >
        w
      </button>
      <button
        value="x"
      >
        x
      </button>
      <button
        value="y"
      >
        y
      </button>
      <button
        value="z"
      >
        z
      </button>
    </p>
    <button>
      Restart Game
    </button>
  </div>
</DocumentFragment>
`;
